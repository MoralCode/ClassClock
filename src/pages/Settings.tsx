import React, { Component } from "react";
import { connect } from "react-redux";
import { push } from "redux-first-routing";
import IPageInterface from "../utils/IPageInterface";
import "../global.css";

export class Settings extends Component<IPageInterface, {}> {
    // document.getElementById("use24HourTime").checked = getLocalStorageBoolean("use24HourTime");

    public updateSettings() {
        // localStorage.setItem("use24HourTime", document.getElementById("use24HourTime").checked);
    }

    // inpit: onchange="updateSettings()"
    render() {
        return (
            <div>
                <a className="navbutton" href="index.html">
                    <i className="fas fa-home" />
                </a>
                <br />
                <section id="options" className="centered topSpace">
                    <label>
                        Use 24-hour Time?
                        <input type="checkbox" id="use24HourTime" />
                    </label>
                    <em>Settings are automatically saved</em>
                </section>
                <section id="credits" className="centered topSpace">
                    <ul className="footer__social">
                        <li>
                            <a href="{{ site.githubURL }}">
                                <i className="fab fa-github" aria-hidden="true" />
                            </a>
                        </li>
                        <li>
                            <a href="{{ site.twitterURL }}">
                                <i className="fab fa-twitter" aria-hidden="true" />
                            </a>
                        </li>
                        <li>
                            <a href="{{ site.instagramURL }}">
                                <i className="fab fa-instagram" aria-hidden="true" />
                            </a>
                        </li>
                    </ul>
                    <p id="credits">
                        Created by:{" "}
                        <a href="https://www.adriancedwards.com">Adrian Edwards</a> and
                        <a href="https://nbdeg.com/">Nick DeGroot</a>
                        <br />
                        Idea by: <a href="https://twitter.com/MrKumprey">Dan Kumprey</a>
                    </p>
                </section>
            </div>
        );
    }
}

export default connect()(Settings);
