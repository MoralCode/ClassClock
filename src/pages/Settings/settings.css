.settingsHeader {
    font-weight: bold;
    text-align: left;
    font-size: 1.2em;
    margin-bottom: 0;
    padding-bottom: 0;
}
