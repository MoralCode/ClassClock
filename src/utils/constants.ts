
export const pages = {
    main: "/",
    fullSchedule: "/schedule",
    settings: "/settings",
    fullScheduleLegacy: "/schedule.html",
    settingsLegacy: "/settings.html",
    selectSchool: "/select",
    loginCallback: "/admin/callback",
    welcome: "/welcome",
    admin: "/admin",
    adminBellSchedule: "/admin/bellschedule"
};

export const URLs = {
    github: "https://github.com/MoralCode/ClassClock",
    slack:
        "https://join.slack.com/t/classclock/shared_invite/enQtNTE0MDkyNzAwNzU3LWNhMGUwODU2ZjhkYTYxMTgzNDE1OWEyMGY2OGNiNTBhOWM5NDVhZGUzNDVlNzRiZTE3NTNmODFjYWNkNDhmMDU",
    twitter: "https://twitter.com/classclockapp",
    instagram: "https://www.instagram.com/classclockapp/",
    bluesky: "https://bsky.app/profile/classclockapp.bsky.social",
    discord: "https://discord.classclock.app"
};
