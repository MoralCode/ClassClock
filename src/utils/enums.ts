export enum TimeStates {
    DAY_OFF = "day off",
    OUTSIDE_SCHOOL_HOURS = "outside school hours",
    SCHOOL_IN_CLASS_OUT = "school is in session, but class is not",
    CLASS_IN_SESSION = "class is in session"
}

export enum TimeComparisons {
    IS_BEFORE = -1, //"before",
    IS_DURING_OR_EXACTLY = 0, //"during/exactly",
    IS_AFTER = 1 //"after"
}
