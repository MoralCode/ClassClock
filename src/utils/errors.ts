// errors.ts
// This file contains classes that represent error conditions
// @author Adrian Edwards



/**
 * This custom Error class is intened to be used for passing a minimum wait 
 * time to the .catch() handlers responsible for retrying requests so requests 
 * arent made during a rate-limit coolown period
 * 
 * This class was created with the help of https://javascript.info/custom-errors
 *
 * @export
 * @class RateLimitError
 * @extends {Error}
 */
export class RateLimitError extends Error {
	wait: number;
	constructor(message:string, msToWait: number) {
		super(message);
		this.name = "RateLimitError";
		this.wait = msToWait;
	}
}