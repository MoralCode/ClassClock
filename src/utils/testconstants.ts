// This is not actually a unit test, just a central place to hold frequently-used constants that are often reused across tests

import Time from "../@types/time";
import School from "../@types/school";
import BellSchedule from "../@types/bellschedule";
import ClassPeriod from "../@types/classperiod";





export const currentDate = new Date("2019-07-28T07:37:50.634Z");

export const className = "First Period";

export const beforeClass = new Time(8, 0);
export const beforeClassString = "8:00"
export const startTime = new Time(8, 25);
export const startTimeString = "8:25";
export const classDuration = new Time(1, 30);
export const duringClass = new Time(9, 0);
export const duringClassString = "9:00";
export const endTime = new Time(9, 55);
export const endTimeString = "9:55";
export const afterClass = new Time(10, 0);
export const afterClassString = "10:00";

export const classPeriod = new ClassPeriod(className, startTime, endTime, currentDate);

export const classPeriodJSON = {
           name: className,
           startTime: startTimeString,
           endTime: endTimeString,
           creationDate: currentDate
       };

export const bellScheduleId = "1";
export const bellScheduleName = "Regular Schedule";
export const bellScheduleEndpoint = "/path/to/sched";
export const bellScheduleDisplayName = "Display Name";


export const bellSchedule = new BellSchedule(
    bellScheduleId,
    bellScheduleName,
    bellScheduleEndpoint,
    [
        new Date("2019-07-28T07:37:50.634Z"),
        new Date("2019-07-29T07:38:10.979Z"),
        new Date("2019-07-23T07:38:28.263Z")
    ],
    [classPeriod],
    currentDate
);

export const bellScheduleJSON = {
           id: bellScheduleId,
           name: bellScheduleName,
           endpoint: bellScheduleEndpoint,
           dates: [
               "2019-07-28T07:37:50.634Z",
               "2019-07-29T07:38:10.979Z",
               "2019-07-23T07:38:28.263Z"
           ],
           classes: [classPeriodJSON],
           lastModified: currentDate
       };


export const schoolId = "10";
export const schoolName = "The High School";
export const schoolAcronym = "THS";
export const schoolEndpoint = "/some/endpoint";
export const schoolTimezone = "America/Cupertino";
export const passingPeriodName = "Transition Time";

export const school = new School(
    schoolId,
    schoolName,
    schoolAcronym,
    schoolEndpoint,
    schoolTimezone,
    [bellSchedule],
    passingPeriodName,
    currentDate,
    currentDate
);

export const schoolJSON = {
           id: schoolId,
           name: schoolName,
           acronym: schoolAcronym,
           endpoint: schoolEndpoint,
           timezone: schoolTimezone,
           schedules: [bellScheduleJSON],
           passingPeriodName,
           creationDate: currentDate,
           lastModified: currentDate
       };
