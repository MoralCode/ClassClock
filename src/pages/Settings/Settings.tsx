import React from "react";
import { connect } from "react-redux";
import { push } from "redux-first-routing";
import "../../global.css";
import "./settings.css";
import { ISchoolsState, SelectedSchoolState } from "../../store/schools/types";
import { ISettingsState, IUserSettings } from "../../store/usersettings/types";
import School from "../../@types/school";
import Link from "../../components/Link";
import { pages } from "../../utils/constants";
import { URLs } from "../../utils/constants";
import { setTimeFormatPreference } from "../../store/usersettings/actions";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faGithub, faTwitter, faInstagram } from "@fortawesome/free-brands-svg-icons";
import { faHome, faSignOutAlt } from "@fortawesome/free-solid-svg-icons";
import { useAuth0 } from "../../react-auth0-wrapper";
import distanceInWords from "date-fns/distance_in_words";
import { selectSchool, invalidateSchool } from "../../store/schools/actions";
import packageJson from '../../package.alias.json';

export interface ISettingProps {
    selectedSchool: SelectedSchoolState;
    userSettings: IUserSettings;
    error: string;
    dispatch: any;
}

const Settings = (props: ISettingProps) => {
    const { logout, user, loginWithRedirect } = useAuth0();

    const navigate = (to: string) => {
        props.dispatch(push(to));
    };

    if (!props.selectedSchool.data && !props.selectedSchool.isFetching) {
        navigate(pages.selectSchool);
    }

    const getAdminButton = () => {
        let isOwner = false;

        let destination;
        let buttonText = isOwner ? "Edit Schedules" : "Log in to edit schedules";
        let hover = "Edit Schedules";

        if (user) {
            isOwner = props.selectedSchool.data.getOwnerIdentifier() === user.sub;
            destination = () => (isOwner ? navigate(pages.admin) : undefined);
            if (!isOwner) {
                hover = "You are not the owner of this class";
            }
        } else {
            destination = () =>
                loginWithRedirect({
                    appState: { targetUrl: pages.admin }
                });

            //TODO: maybe have it automatically find the school you ARE an admin of
        }

        return (
            <Link
                // tslint:disable-next-line: jsx-no-lambda
                destination={destination}
                title={hover}
            >
                <button disabled={undefined}>{buttonText}</button>
            </Link>
        );
    };

    const getVersionHTML = () => {
        if (process.env.REACT_APP_VERCEL_GIT_COMMIT_SHA) {

            
            const version_sha = process.env.REACT_APP_VERCEL_GIT_COMMIT_SHA.substring(0, 6)

            const githubURL = "https://" + packageJson.repository.replace(":", ".com/") + "/commit/"
            
            return <>(<a href={githubURL + version_sha}>{version_sha}</a>)</>
        } else {
            return <></>
        }
    }

    const selectedSchoolInfo = () => {
        if (props.selectedSchool.isFetching) {
            return <span>Loading School...</span>
        } else if (props.error) {
            return <span>An Error Occurred</span>
        } else {
            return <>
                <Link
                    // tslint:disable-next-line: jsx-no-lambda
                    destination={() => navigate(pages.selectSchool)}
                    title="Change School"
                >
                    {props.selectedSchool.data.getName() + " "}
                </Link>
                <br />
                <em className="smallerText">
                    Last updated{" "}
                    {distanceInWords(
                        new Date(),
                        new Date(props.selectedSchool.lastUpdated)
                    ) + " ago "}
                </em>

                <em className="smallerText">
                    <Link
                        // tslint:disable-next-line: jsx-no-lambda
                        destination={() =>
                            props.dispatch(
                                selectSchool(
                                    props.selectedSchool.data.getIdentifier()
                                )
                            )
                        }
                        title="Reload Schedule"
                    >
                        Refresh
                        </Link>
                </em>
            </>
        }
    }

    return (
        <div>
            <Link
                className="cornerNavButton cornerNavTop cornerNavLeft smallIcon"
                // tslint:disable-next-line: jsx-no-lambda
                destination={() => navigate(pages.main)}
            >
                <FontAwesomeIcon icon={faHome} />
            </Link>
            {user && (
                <Link
                    className="cornerNavButton cornerNavTop cornerNavRight smallIcon"
                    // tslint:disable-next-line: jsx-no-lambda
                    destination={() => {
                        logout();
                    }}
                    title="Log Out"
                >
                    <FontAwesomeIcon icon={faSignOutAlt} />
                </Link>
            )}

            <h1>Settings</h1>
            <h2 className="settingsHeader centeredWidth">Selected School: </h2>
            <div>
                {selectedSchoolInfo()}
                <br />
                <br />
                {getAdminButton()}
            </div>
            <h2 className="settingsHeader centeredWidth">Time Display:</h2>
            <label>
                Use 24-hour Time?{" "}
                <input
                    type="checkbox"
                    checked={props.userSettings.use24HourTime}
                    // tslint:disable-next-line: jsx-no-lambda
                    onChange={() => {
                        props.dispatch(
                            setTimeFormatPreference(!props.userSettings.use24HourTime)
                        );
                    }}
                />
            </label>
            <br />
            <p>
                <em className="smallerText">
                    Settings are automatically saved on your device
                </em>
            </p>

            <p style={{ marginTop: "20vh" }}>Follow ClassClock:</p>
            <ul className="footer__social">
                <li>
                    <Link destination={URLs.github}>
                        <FontAwesomeIcon icon={faGithub} />
                    </Link>
                </li>
                <li>
                    <Link destination={URLs.twitter}>
                        <FontAwesomeIcon icon={faTwitter} />
                    </Link>
                </li>
                <li>
                    <Link destination={URLs.instagram}>
                        <FontAwesomeIcon icon={faInstagram} />
                    </Link>
                </li>
            </ul>
            <p id="credits">
                Created by: <a href={process.env.REACT_APP_AUTHORURL}>Adrian Edwards</a>{" "}
                and <a href="https://nickthegroot.com/">Nick DeGroot</a>
                <br />
                Idea by: <a href="https://twitter.com/MrKumprey">Dan Kumprey</a>
            </p>
            <p style={{ fontSize: "smaller" }}>ClassClock version {packageJson.version} {getVersionHTML()}</p>
        </div>
    );
};

const mapStateToProps = (state: ISchoolsState & ISettingsState & {error: string}) => {
    const { selectedSchool, userSettings, error } = state;
    return {
        selectedSchool: Object.assign({}, selectedSchool, {
            data: School.fromJson(selectedSchool.data)
        }),
        userSettings,
        error
    };
};

export default connect(mapStateToProps)(Settings);
