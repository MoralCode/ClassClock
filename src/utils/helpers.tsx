import Time from "../@types/time";
import School from "../@types/school";
import { TimeComparisons, TimeStates } from "./enums";
import ClassPeriod from "../@types/classperiod";

export function getValueIfKeyInList(list: string[], object: any) {
    for (const key of list) {
        if (object.hasOwnProperty(key)) {
            return object[key];
        }
    }
}

export function deconstructJsonApiResource(json: any) {
    const data = {
        type: json.type,
        id: json.id,
        ...(json.links !== undefined && {
            endpoint: json.links.self
        })
    };
    return Object.assign({}, data, json.attributes);
}

export function getCurrentDate() {
    return new Date();
}

/**
 * @returns a flag that represents the current chunk of time categorically
 */
export function getTimeStateForDateAtSchool(date: Date, school: School) {
    const currentBellSchedule = school.getScheduleForDate(date);

    //there is no schedule that applies today
    if (!currentBellSchedule) {
        return TimeStates.DAY_OFF;
    }

    const currentClassPeriod = currentBellSchedule.getClassPeriodForTime(
        Time.fromDate(date)
    );

    //it is a school day but it is not school hours
    if (!school.isInSession(date)) {
        return TimeStates.OUTSIDE_SCHOOL_HOURS;
    }

    //the current time lies between the start of the first schedules class and the end of the last
    else if (school.isInSession(date) && !currentClassPeriod) {
        return TimeStates.SCHOOL_IN_CLASS_OUT;
    }

    //the current time lies within a scheduled class period
    else if (currentClassPeriod !== undefined) {
        return TimeStates.CLASS_IN_SESSION;
    }
}

/**
 * @returns the next relevent time to count down to
 */
export function getNextImportantTime(
    date: Date,
    school: School
): [ClassPeriod, Time] | undefined {
    const currentBellSchedule = school.getScheduleForDate(date);

    //there is no schedule that applies today
    if (!currentBellSchedule) {
        return;
    }

    currentBellSchedule.getAllClasses().forEach((value: ClassPeriod) => {
        for (const time of [value.getStartTime(), value.getEndTime()]) {
            if (time.getMillisecondsTo(Time.fromDate(date)) >= 0) {
                return [value, time];
            }
        }
    });
}

/**
 * This export function checks if the current time is between the two given times
 * This is useful for checking which class period you are currently in or for checking if school is in session.
 *
 * @param {*} checkTime the time that the check results are returned for
 * @param {*} startTime the start time of the range to check
 * @param {*} endTimethe the end time of the range to check
 *
 * @returns -1 if checkTime is before range, 0 if checkTime is within range, 1 if checkTime is after range
 */
export function checkTimeRange(checkTime: Time, startTime: Time, endTime: Time) {
    if (startTime.getMillisecondsTo(endTime) <= 0) {
        //theres a problem
    }
    const startCheck = checkTime.getMillisecondsTo(startTime);
    const endCheck = checkTime.getMillisecondsTo(endTime);

    if (startCheck > 0) {
        return TimeComparisons.IS_BEFORE;
    } else if (endCheck < 0) {
        return TimeComparisons.IS_AFTER;
    } else {
        return TimeComparisons.IS_DURING_OR_EXACTLY;
    }
}
