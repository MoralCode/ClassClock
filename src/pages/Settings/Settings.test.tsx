import React from 'react';
import renderer from "react-test-renderer";
import { school } from '../../utils/testconstants';
import { Settings } from './Settings';


// make a fake dispatch function, it doesnt need to do anything, because we assume that has already been tested in the redux library
const mockDispatch = jest.fn();

describe("Settings", () => {

  it("renders correctly", () => {
    const tree = renderer.create(<Settings dispatch={mockDispatch} selectedSchool={{
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1659889423,
      data: school
    }} userSettings={{use24HourTime: true}} error={''} />).toJSON();
      expect(tree).toMatchSnapshot();
  });
});


