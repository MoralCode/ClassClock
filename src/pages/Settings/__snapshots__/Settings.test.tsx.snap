// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings renders correctly 1`] = `
<div>
  <a
    className="cornerNavButton cornerNavTop cornerNavLeft smallIcon"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-house "
      data-icon="house"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 576 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M575.8 255.5C575.8 273.5 560.8 287.6 543.8 287.6H511.8L512.5 447.7C512.5 450.5 512.3 453.1 512 455.8V472C512 494.1 494.1 512 472 512H456C454.9 512 453.8 511.1 452.7 511.9C451.3 511.1 449.9 512 448.5 512H392C369.9 512 352 494.1 352 472V384C352 366.3 337.7 352 320 352H256C238.3 352 224 366.3 224 384V472C224 494.1 206.1 512 184 512H128.1C126.6 512 125.1 511.9 123.6 511.8C122.4 511.9 121.2 512 120 512H104C81.91 512 64 494.1 64 472V360C64 359.1 64.03 358.1 64.09 357.2V287.6H32.05C14.02 287.6 0 273.5 0 255.5C0 246.5 3.004 238.5 10.01 231.5L266.4 8.016C273.4 1.002 281.4 0 288.4 0C295.4 0 303.4 2.004 309.5 7.014L564.8 231.5C572.8 238.5 576.9 246.5 575.8 255.5L575.8 255.5z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </a>
  <h1>
    Settings
  </h1>
  <h2
    className="settingsHeader centeredWidth"
  >
    Selected School: 
  </h2>
  <div>
    <a
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
      title="Change School"
    >
      The High School 
    </a>
    <br />
    <em
      className="smallerText"
    >
      Last updated
       
      over 54 years ago 
    </em>
    <em
      className="smallerText"
    >
      <a
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        title="Reload Schedule"
      >
        Refresh
      </a>
    </em>
    <br />
    <br />
  </div>
  <h2
    className="settingsHeader centeredWidth"
  >
    Time Display:
  </h2>
  <label>
    Use 24-hour Time?
     
    <input
      checked={true}
      onChange={[Function]}
      type="checkbox"
    />
  </label>
  <br />
  <p>
    <em
      className="smallerText"
    >
      Settings are automatically saved on your device
    </em>
  </p>
  <p
    style={
      Object {
        "marginTop": "20vh",
      }
    }
  >
    Follow ClassClock:
  </p>
  <ul
    className="footer__social"
  >
    <li>
      <a
        href="https://github.com/MoralCode/ClassClock"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-github "
          data-icon="github"
          data-prefix="fab"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 496 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3 .3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5 .3-6.2 2.3zm44.2-1.7c-2.9 .7-4.9 2.6-4.6 4.9 .3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3 .7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3 .3 2.9 2.3 3.9 1.6 1 3.6 .7 4.3-.7 .7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3 .7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3 .7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </a>
    </li>
    <li>
      <a
        href="https://twitter.com/classclockapp"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-twitter "
          data-icon="twitter"
          data-prefix="fab"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M459.4 151.7c.325 4.548 .325 9.097 .325 13.65 0 138.7-105.6 298.6-298.6 298.6-59.45 0-114.7-17.22-161.1-47.11 8.447 .974 16.57 1.299 25.34 1.299 49.06 0 94.21-16.57 130.3-44.83-46.13-.975-84.79-31.19-98.11-72.77 6.498 .974 12.99 1.624 19.82 1.624 9.421 0 18.84-1.3 27.61-3.573-48.08-9.747-84.14-51.98-84.14-102.1v-1.299c13.97 7.797 30.21 12.67 47.43 13.32-28.26-18.84-46.78-51.01-46.78-87.39 0-19.49 5.197-37.36 14.29-52.95 51.65 63.67 129.3 105.3 216.4 109.8-1.624-7.797-2.599-15.92-2.599-24.04 0-57.83 46.78-104.9 104.9-104.9 30.21 0 57.5 12.67 76.67 33.14 23.72-4.548 46.46-13.32 66.6-25.34-7.798 24.37-24.37 44.83-46.13 57.83 21.12-2.273 41.58-8.122 60.43-16.24-14.29 20.79-32.16 39.31-52.63 54.25z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </a>
    </li>
    <li>
      <a
        href="https://www.instagram.com/classclockapp/"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-instagram "
          data-icon="instagram"
          data-prefix="fab"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.1 141c-63.6 0-114.9 51.3-114.9 114.9s51.3 114.9 114.9 114.9S339 319.5 339 255.9 287.7 141 224.1 141zm0 189.6c-41.1 0-74.7-33.5-74.7-74.7s33.5-74.7 74.7-74.7 74.7 33.5 74.7 74.7-33.6 74.7-74.7 74.7zm146.4-194.3c0 14.9-12 26.8-26.8 26.8-14.9 0-26.8-12-26.8-26.8s12-26.8 26.8-26.8 26.8 12 26.8 26.8zm76.1 27.2c-1.7-35.9-9.9-67.7-36.2-93.9-26.2-26.2-58-34.4-93.9-36.2-37-2.1-147.9-2.1-184.9 0-35.8 1.7-67.6 9.9-93.9 36.1s-34.4 58-36.2 93.9c-2.1 37-2.1 147.9 0 184.9 1.7 35.9 9.9 67.7 36.2 93.9s58 34.4 93.9 36.2c37 2.1 147.9 2.1 184.9 0 35.9-1.7 67.7-9.9 93.9-36.2 26.2-26.2 34.4-58 36.2-93.9 2.1-37 2.1-147.8 0-184.8zM398.8 388c-7.8 19.6-22.9 34.7-42.6 42.6-29.5 11.7-99.5 9-132.1 9s-102.7 2.6-132.1-9c-19.6-7.8-34.7-22.9-42.6-42.6-11.7-29.5-9-99.5-9-132.1s-2.6-102.7 9-132.1c7.8-19.6 22.9-34.7 42.6-42.6 29.5-11.7 99.5-9 132.1-9s102.7-2.6 132.1 9c19.6 7.8 34.7 22.9 42.6 42.6 11.7 29.5 9 99.5 9 132.1s2.7 102.7-9 132.1z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </a>
    </li>
    <li>
      <a
        href="https://discord.classclock.app"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-discord "
          data-icon="discord"
          data-prefix="fab"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M524.5 69.84a1.5 1.5 0 0 0 -.764-.7A485.1 485.1 0 0 0 404.1 32.03a1.816 1.816 0 0 0 -1.923 .91 337.5 337.5 0 0 0 -14.9 30.6 447.8 447.8 0 0 0 -134.4 0 309.5 309.5 0 0 0 -15.14-30.6 1.89 1.89 0 0 0 -1.924-.91A483.7 483.7 0 0 0 116.1 69.14a1.712 1.712 0 0 0 -.788 .676C39.07 183.7 18.19 294.7 28.43 404.4a2.016 2.016 0 0 0 .765 1.375A487.7 487.7 0 0 0 176 479.9a1.9 1.9 0 0 0 2.063-.676A348.2 348.2 0 0 0 208.1 430.4a1.86 1.86 0 0 0 -1.019-2.588 321.2 321.2 0 0 1 -45.87-21.85 1.885 1.885 0 0 1 -.185-3.126c3.082-2.309 6.166-4.711 9.109-7.137a1.819 1.819 0 0 1 1.9-.256c96.23 43.92 200.4 43.92 295.5 0a1.812 1.812 0 0 1 1.924 .233c2.944 2.426 6.027 4.851 9.132 7.16a1.884 1.884 0 0 1 -.162 3.126 301.4 301.4 0 0 1 -45.89 21.83 1.875 1.875 0 0 0 -1 2.611 391.1 391.1 0 0 0 30.01 48.81 1.864 1.864 0 0 0 2.063 .7A486 486 0 0 0 610.7 405.7a1.882 1.882 0 0 0 .765-1.352C623.7 277.6 590.9 167.5 524.5 69.84zM222.5 337.6c-28.97 0-52.84-26.59-52.84-59.24S193.1 219.1 222.5 219.1c29.67 0 53.31 26.82 52.84 59.24C275.3 310.1 251.9 337.6 222.5 337.6zm195.4 0c-28.97 0-52.84-26.59-52.84-59.24S388.4 219.1 417.9 219.1c29.67 0 53.31 26.82 52.84 59.24C470.7 310.1 447.5 337.6 417.9 337.6z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </a>
    </li>
  </ul>
  <p
    id="credits"
  >
    Created by: 
    <a
      href="https://www.adriancedwards.com"
    >
      Adrian Edwards
    </a>
     
    and 
    <a
      href="https://nickthegroot.com/"
    >
      Nick DeGroot
    </a>
    <br />
    Idea by: 
    <a
      href="https://twitter.com/MrKumprey"
    >
      Dan Kumprey
    </a>
  </p>
  <p
    style={
      Object {
        "fontSize": "smaller",
      }
    }
  >
    ClassClock version 
    0.4.1
     
  </p>
</div>
`;
